### PR TITLE
Check if conda dep in noarch filter

### LINKED ISSUE #264: No dependency of conda for noarch
I think we may have the information needed to automagic this one.

### PR DESCRIPTION
**Pull request by justcalamari**
Closes #264

### CODE REVIEW

**Comment by CJ-Wright** [comment]
This is having some issues with the tests, would it be possible to write dummy graph for the tests?

File: conda_forge_tick/migrators.xsh [hunk 1]
```diff
@@ -423,6 +423,11 @@ class Noarch(Migrator):
         for line in attrs.get('raw_meta_yaml', '').splitlines():
             if self.sel_pat.match(line):
                 return True
+
+        # Not a dependency of `conda`
+        if attrs['feedstock_name'] in $GRAPH.node['conda']['reqs']:
```
Comment (isuruf) [inline conda_forge_tick/migrators.xsh]: You might need to traverse the graph since some deps might be transitive.

**Comment by justcalamari** [comment]
@CJ-Wright, @scopatz, @isuruf Ready for review

**Comment by CJ-Wright** [comment]
Awesome!

**Comment by CJ-Wright** [comment]
Thank you very much @justcalamari

**Comment by jakirkham** [comment]
Thanks for this. :)

Is it possible to get a listing of `conda`'s existing dependencies (given some may be indirect)? That way we can check for other PRs that might have already been merged before this change.

**Comment by CJ-Wright** [comment]
```python
import networkx as nx
g = nx.read_gpickle('graph.pkl')
nx.ancestors(g, 'conda')
```
gave me
```python
{'asn1crypto',
 'atomicwrites',
 'attrs',
 'autoconf',
 'automake',
 'bison',
 'bzip2',
 'ca-certificates',
 'certifi',
 'cffi',
 'chardet',
 'cmake',
 'colorama',
 'conda-env',
 'cryptography',
 'curl',
 'cython',
 'enum34',
 'expat',
 'flex',
 'funcsigs',
 'idna',
 'ipaddress',
 'krb5',
 'libffi',
 'libiconv',
 'libssh2',
 'libtool',
 'libuv',
 'm4',
 'menuinst',
 'more-itertools',
 'msinttypes',
 'ncurses',
 'openssl',
 'perl',
 'pip',
 'pkg-config',
 'pluggy',
 'py',
 'pycosat',
 'pycparser',
 'pyopenssl',
 'pysocks',
 'pytest',
 'pytest-runner',
 'python',
 'pywin32',
 'readline',
 'requests',
 'rhash',
 'ruamel_yaml',
 'setuptools',
 'setuptools_scm',
 'six',
 'sqlite',
 'tk',
 'toolchain',
 'urllib3',
 'wheel',
 'win_inet_pton',
 'wincertstore',
 'xz',
 'yaml',
 'zlib'}
```
(Note that I'm working off an old copy of the graph)

I may also be possible to use the `PRed_json` data in each of those nodes to close any open `noarch` PRs.

**Comment by isuruf** [comment]
[This](https://travis-ci.org/conda-forge/conda-feedstock/jobs/402262467#L964-L992) shows that `asn1crypto, chardet, idna, ipaddress, pycparser` are already noarch.

**Comment by jakirkham** [comment]
Thanks.

Good point. Most of those are recent changes with the exception of `ipaddress`. Interesting that the latter hasn't caused us problems.

**Comment by jakirkham** [comment]
Alright have triaged the PRs. Listed above thanks to the xrefs. Sadly many more have been merged than we would like. In a few cases conversion to `noarch: python` happened a while ago. Not sure why they didn't cause issues as well. Maybe they are build only dependencies of conda?

**Comment by CJ-Wright** [comment]
Thank you very much! I'm very sorry those PRs were sent, although checking for deps of conda were part of the checklist it seems that this could have used some more automagic.

Would it be possible to add members of core to those feedstocks? They seem of systemic importance.

**Comment by jakirkham** [comment]
Yeah, unlike some of the other things, this is a hard thing to check. Further if `conda` adds new dependencies, we can end up already having this problem without knowing. We probably need a few checks including the migration bot, linter, and human eyes.

Agree. Had added some to `conda` and such in the early days of conda-forge. We probably need some more though.

**Comment by jakirkham** [comment]
Should we try to automate reversion PRs or are we requesting maintainers to do this? Question came to me through a side channel from a maintainer.

**Comment by CJ-Wright** [comment]
I guess that depends on how many PRs there are too revert. My current opinion is that core should roll them back since this is a system issue.

**Comment by CJ-Wright** [comment]
Although maybe we should write a migrator since this issue will crop up again and again.  Would it be possible to add something to the conda recipe to denote that it's dependencies can't be noarch? Are there other feedstocks like that?

**Comment by isuruf** [comment]
Btw, are there any people using conda 4.2 these days? If everybody is on 4.3, this is not an issue

**Comment by jakirkham** [comment]
We’ve been getting reports of issues with `noarch: python` packages that `conda` depends on getting installed incorrectly when using `conda` 4.5.6. So unfortunately having a newer version of `conda` does not seem sufficient for solving the problem.

**Comment by CJ-Wright** [comment]
@jakirkham should we add to those recipes the thing that tests the new recipes against downstream packages? This could allow us to make certain conda doesn't break, maybe?

**Comment by mbargull** [comment]
> We’ve been getting reports of issues with `noarch: python` packages that `conda` depends on getting installed incorrectly when using `conda 4.5.6`. So unfortunately having a newer version of `conda` does not seem sufficient for solving the problem.

Can point to any case that uses `conda >=4.3` and has problems installing `noarch: python`? I have a hard time seeing how that'd happen.

> @jakirkham  should we add to those recipes the thing that tests the new recipes  against downstream packages? This could allow us to make certain conda  doesn't break, maybe?

Since I don't think `conda >=4.3` is affected, you'd have to downstream test `conda <4.3.0a0`.

**Comment by CJ-Wright** [comment]
> Since I don't think `conda >=4.3` is affected, you'd have to downstream test `conda <4.3.0a0`.

Hmm that is a fair point, I guess we should decide what versions of conda we will actively test against.

I also meant that more as a systemic statement. We may need to come up with a list of recipes which are "too big to fail" because of their importance to the ecosystem. To ensure the stability of those recipes we might need to:
1. Have at least one member of core on the maintainers list (bonus points if we can get all core on them)
1. Have the packages which the systemic packages depend on test to make sure they don't break the systemic packages when they are updated.
1. Maybe put in review requirements? (This may slow things down but may also be needed?)
1. More checks that I haven't thought of.

The main point is that if this move to noarch happend prior to `conda 4.3` we would have most likely crippled the entire ecosystem. Any checks that we can have (especially if they are automated) to guard against that may be good to have.

I'm going to xref this ton conda-forge.io so I can ping core.

**Pull request merged**